### PR TITLE
Make retryable errors configurable

### PR DESCRIPTION
**Pull request by infraredgirl**
### Description

This change enables users to create their own custom list of retryable errors in the `terragrunt.hcl`.

### Related issue
Closes: #1040

### How to test

To test this you can use these 2 files:
`main.tf`:

```tcl
resource "null_resource" "example" {
  triggers = {
    always_recreate = timestamp()
  }

provisioner "local-exec" {
    command = "echo 'My own little error' && exit 1"
  }
}

output "text" {
  value = "Hello, World"
}
```

`terragrunt.hcl`:
```tcl
retryable_errors = [
"My own little error"
]
```

Then just run `go run main.go apply` and the error which is specified in the config file should be detected as retryable and retries attempted. To make sure other errors are not picked up as retryable, edit the `echo` statement in `main.tf` to output a different error message, and re-run `go run main.go apply`. Now `terragrunt` should just fail without attempting retries.

### Known issues / still TBD:

- [x] Add unit tests
- [x] Add integration tests
~- [ ] Change the format of the retryable errors structure. Right now it's a list of regex strings and we want a map of regexes and error descriptions (as described in the [issue](#1040)).~ We decided to stick with using the list for now.
- [x] Make the config option optional (right now terragrunt errors out if this option is not set in the config file; it should fall back to using the default hardcoded value instead)

### CODE REVIEW

File: config/config_test.go
Comment (brikis98) [inline config/config_test.go:173]: NIT: `require.NoError(t, err)` is slightly shorter/simpler.

File: config/config_test.go
Comment (brikis98) [inline config/config_test.go:181]: NIT: I think the following will do the same thing, but shorter, and with clearer error reporting:

```suggestion
		assert.Equal(t, []string{"My own little error", "Another one of my errors"}, terragruntConfig.RetryableErrors)
```

File: config/config_test.go
Comment (brikis98) [inline config/config_test.go:185]: Same comments as above

File: config/config_test.go
Comment (brikis98) [inline config/config_test.go:192]: Consider a list of 1 item or 3 items just so it's not exactly the same as the HCL version and exercises a few more code paths

File: config/config_test.go
Comment (brikis98) [inline config/config_test.go:161]: Consider updating some of the other existing test cases to check that `terragruntConfig.RetryableErrors` is `nil`.

**Comment by infraredgirl** [comment]
Thanks so much - updated the PR to address the comments.

Last remaining item is integration test - looking into that now!

File: test/integration_test.go
Comment (brikis98) [inline test/integration_test.go:1184]: This checks that the `apply` eventually failed, but how do we know this actually did some retries? 

You probably want to test a fixture that fails the first N items but then succeeds on try N+1. Here's one approach for doing that:

1. Update the Terraform test fixture to:

    1. Take in a `url` variable.
    1. Use the [`http`](https://www.terraform.io/docs/providers/http/data_source.html) data source to make an HTTP request to `url`.
    1. Update the Bash script in the `null_resource` to exit with an error if the response the `http` data source gets is not a 200 OK.
    1.  Add `triggers { run_always = timestamp() }` to the `null_resource` so it runs every time you run `apply`, and not just on the very first time.

1. Update the integration test to:

    1. Fire up a [local HTTP server](https://golang.org/pkg/net/http/) listening on some port.
    1. Configure that HTTP server to return a 500 error on the first N requests, but a 200 OK on request N+1.
    1. Run `terragrunt apply -var url=http://localhost:<PORT>`, make sure the logs mention errors/retries, but that the `apply` ultimately succeeds.

Let me know if that makes sense!

Comment (infraredgirl) [reply test/integration_test.go:1184]: It makes sense but I wonder if firing up an HTTP server is a bit of an overkill - maybe we can do it by keeping the state in a file in `/tmp` (like we discussed this morning). What do you think?

Comment (brikis98) [reply test/integration_test.go:1184]: Yup, you could use a file-based approach too, but it might be a bit more error prone as files have to be cleaned up so that the outputs from previous test runs don't cause issues with subsequent test runs. To do it with files, you'd do something like this:

1. Use the [`random_id`](https://www.terraform.io/docs/providers/random/r/id.html) resource to generate a unique filename on the very first `apply`.

1. Update the Bash script in the `null_resource` to: 

    1. Check if the file with the name from `random_id` exists in `/tmp`. 
    1. If it does not, write it to disk, and exit with an error. 
    1. If it does, exit without an error.

1. Add `triggers { run_always = timestamp() }` to the `null_resource` so it runs every time you run `apply`, and not just on the very first time.

This way, the module will fail the first time you run `apply`, but succeed after that.

File: cli/cli_app.go
Comment (yorinasub17) [inline cli/cli_app.go:324]: Not for this PR, but we should probably expose a helper function that returns all the default retryable errors. I think 90% of use cases for this will want to append to the default list, rather than replace. If we had a helper (e.g. `get_default_retryable_errors`), then one could do the append use case with:

```
retryable_errors = concat(get_default_retryable_errors(), ["my special snowflake"])
```

File: config/config_as_cty.go
Comment (yorinasub17) [inline config/config_as_cty.go:71]: This was at first confusing to me because `RetryableErrors` is not a struct, but then I realized that `ImpliedType` "does the right thing." We should probably rename `gostructToCty` eventually...

**Review by yorinasub17 — DISMISSED** [review]
Overall implementation LGTM! Only had 2 minor comments for future improvements.

File: test/integration_test.go
Comment (infraredgirl) [reply test/integration_test.go]: Awesome - done!

File: cli/cli_app.go
Comment (infraredgirl) [reply cli/cli_app.go:324]: Agreed, that makes total sense.

File: config/config_as_cty.go
Comment (infraredgirl) [reply config/config_as_cty.go:71]: I went through the very same thought process... :)

File: test/fixture-auto-retry/custom-errors/script.sh
Comment (brikis98) [inline test/fixture-auto-retry/custom-errors/script.sh:7]: Oh, haha, cleaning up the file on success is clever 😁

File: test/fixture-auto-retry/custom-errors/script.sh
Comment (brikis98) [inline test/fixture-auto-retry/custom-errors/script.sh:3]: NIT: we mostly follow the [Google shell style guide](https://google.github.io/styleguide/shellguide.html). Same for a few other suggestions below.

```suggestion
FILENAME="/tmp/$1"
```

File: test/fixture-auto-retry/custom-errors/script.sh
Comment (brikis98) [inline test/fixture-auto-retry/custom-errors/script.sh:7]:
```suggestion
  rm "$FILENAME"
```

File: test/fixture-auto-retry/custom-errors/script.sh
Comment (brikis98) [inline test/fixture-auto-retry/custom-errors/script.sh:10]:
```suggestion
  touch "$FILENAME"
```

File: test/integration_test.go
Comment (brikis98) [inline test/integration_test.go:1174]: Prob worth adding another test that uses the same module, but does _not_ set retryable errors to make sure that test fails!

Comment (infraredgirl) [reply test/integration_test.go:1174]: Addressed everything but this one - added a TODO as discussed on Slack

**Comment by infraredgirl** [comment]
Marking as WIP while still working on this.

Issues left to address:

- [ ] Rebase and fix conflicts
- [ ] Add docs from PR #1231 to this PR and close the other PR
- [ ] Add the missing integration test (see [this comment](https://github.com/gruntwork-io/terragrunt/pull/1232#discussion_r444320783))

File: test/integration_test.go
Comment (infraredgirl) [reply test/integration_test.go:1203]: This is now addressed in 7dfdc8c.

File: cli/cli_app.go
Comment (brikis98) [reply cli/cli_app.go:331]: @infraredgirl Could you file an issue to track this?

File: config/config_as_cty.go
Comment (brikis98) [reply config/config_as_cty.go:74]: Might want a GitHub issue for this too.

**Review by brikis98 — APPROVED** [review]
LGTM! I think this is ready to merge if the tests are passing :)

**Comment by brikis98** [comment]
@infraredgirl Feel free to merge, do a release, and add to our newsletter!

File: cli/cli_app.go
Comment (infraredgirl) [reply cli/cli_app.go:331]: Done: https://github.com/gruntwork-io/terragrunt/issues/1383

File: config/config_as_cty.go
Comment (infraredgirl) [reply config/config_as_cty.go:74]: Done: https://github.com/gruntwork-io/terragrunt/issues/1384

**Comment by infraredgirl** [comment]
> LGTM! I think this is ready to merge if the tests are passing :)

Thanks for the review @brikis98! Tests are passing for me locally. `terragrunt` doesn't seem to have any automated tests/CI hooked up to run on PR submissions AFAICS, so are we good to merge when tests pass locally?

**Comment by brikis98** [comment]
Ohhh, it does have tests, but they don't run for PRs submitted by people outside our org. It's weird that it doesn't update that after you've joined the org... Perhaps some GitHub bug/limit?

Might be a good chance to practice with this tool that we use to trigger test runs for PRs submitted by people outside the org: https://github.com/gruntwork-io/prototypes/tree/master/testpr

**Comment by infraredgirl** [comment]
Ah, it's probably because the initial commits in this PR were submitted with my old/non-Gruntwork email. Thanks for the pointer - kicked off the CI build now. Looks like we have some tests failures - looking into those now.

**Comment by infraredgirl** [comment]
Unit tests are now passing however integration tests seem to be failing on something unrelated to this PR. I can see from previous PR runs that the same 16 integration tests are failing starting 2 days ago. The errors are all the same: `Failed to list object versions in s3 bucket terragrunt-test-bucket-djstsa: NoSuchBucket: The specified bucket does not exist`. Any ideas about this?

**Comment by brikis98** [comment]
> Unit tests are now passing however integration tests seem to be failing on something unrelated to this PR. I can see from previous CI runs that the same 16 integration tests are failing starting 2 days ago. The errors are all the same: `Failed to list object versions in s3 bucket terragrunt-test-bucket-djstsa: NoSuchBucket: The specified bucket does not exist`. Any ideas about this?

Might be a bit of a red herring... 

Do the tests pass when you run one of them at a time locally? If not, could you share the full log output?

**Comment by infraredgirl** [comment]
Alright, after a few yak shaves (thanks @brikis98 for the help debugging) we have a passing test!

Can I have a final approval please? :)

**Review by brikis98 — APPROVED** [review]
Ship it! 🎉

**Comment by infraredgirl** [comment]
Merged, released and updated newsletter!

**Comment by Fran-ELS-AMS** [comment]
Is there currently any way to add `retryable_errors` to the default hard-coded list, instead of replacing the whole list?

Currently we're copying the whole default list from here and adding our custom errors, but if the default list gets updated we have to manually update our code

**Comment by infraredgirl** [comment]
> Is there currently any way to add `retryable_errors` to the default hard-coded list, instead of replacing the whole list?
> 
> Currently we're copying the whole default list from here and adding our custom errors, but if the default list gets updated we have to manually update our code

This is not currently supported but we do plan to add it in the future - in fact we already have an open issue for this very feature: https://github.com/gruntwork-io/terragrunt/issues/1383.

**Comment by Fran-ELS-AMS** [comment]
> > Is there currently any way to add `retryable_errors` to the default hard-coded list, instead of replacing the whole list?
> > Currently we're copying the whole default list from here and adding our custom errors, but if the default list gets updated we have to manually update our code
> 
> This is not currently supported but we do plan to add it in the future - in fact we already have an open issue for this very feature: #1383.

Great, thanks a lot! 🙇

**Comment by jubr** [comment]
This cool new feature got me 🤔: could a similar approach also work for the
```
Error: Could not load plugin
Plugin reinitialization required. Please run "terraform init".
```
use case? I'm running into this quite regularly in my dev workflow.

**Comment by infraredgirl** [comment]
> This cool new feature got me 🤔: could a similar approach also work for the
> 
> ```
> Error: Could not load plugin
> Plugin reinitialization required. Please run "terraform init".
> ```
> 
> use case? I'm running into this quite regularly in my dev workflow.

In other words, terragrunt should automatically run `init` when needed. I think it's worth considering - @jubr would you mind opening an issue for this and we can take it from there?

**Comment by jubr** [comment]
@infraredgirl ✅ #1414

**Pull request merged**
